### PR TITLE
Fixed pokemon & kirby link communication errors

### PR DESCRIPTION
**Pull request by kakashidinho**
Basically, let slaves wait until they use all added cycles from `addCycles` callback before signalling master to continue.

Also added macro switch to allow switching between using mutex vs using atomic ops for lockstep.

### CODE REVIEW

File: include/mgba-util/common.h [hunk 1]
```diff
@@ -77,14 +77,24 @@ typedef intptr_t ssize_t;
 #define ATOMIC_STORE(DST, SRC) __atomic_store_n(&DST, SRC, __ATOMIC_RELEASE)
 #define ATOMIC_LOAD(DST, SRC) DST = __atomic_load_n(&SRC, __ATOMIC_ACQUIRE)
 #define ATOMIC_ADD(DST, OP) __atomic_add_fetch(&DST, OP, __ATOMIC_RELEASE)
+#define ATOMIC_SUB(DST, OP) __atomic_sub_fetch(&DST, OP, __ATOMIC_RELEASE)
```
Comment (endrift) [inline include/mgba-util/common.h]: There was no `ATOMIC_SUB` for a reason: you can just negate the operand.

File: include/mgba-util/platform/posix/threading.h [hunk 2]
```diff
@@ -29,6 +29,18 @@ static inline int MutexInit(Mutex* mutex) {
 	return pthread_mutex_init(mutex, 0);
 }
 
+static inline int MutexInitRecursive(Mutex* mutex) {
```
Comment (endrift) [inline include/mgba-util/platform/posix/threading.h]: Why did you add this? You'd need to port it to all of the threading platforms and I don't think it's necessary.

File: include/mgba/core/lockstep.h [hunk 3]
```diff
@@ -19,14 +22,20 @@ enum mLockstepPhase {
...
 
 struct mLockstep {
+#if MGBA_LOCK_STEP_USE_MUTEX
+	Mutex mutex;
+	Condition cond;
+#endif
+
 	int attached;
-	enum mLockstepPhase transferActive;
+	int32_t transferActive; // must be one of the values in mLockstepPhase
```
Comment (endrift) [inline include/mgba/core/lockstep.h]: Why did you change this away from an enum?

File: include/mgba/internal/gb/sio/lockstep.h [hunk 4]
```diff
@@ -38,6 +38,7 @@ struct GBSIOLockstepNode {
 };
 
 void GBSIOLockstepInit(struct GBSIOLockstep*);
+void GBSIOLockstepInit2(struct GBSIOLockstep*, bool initBase);
```
Comment (endrift) [inline include/mgba/internal/gb/sio/lockstep.h]: Why is this needed?

File: src/gb/sio/lockstep.c [hunk 5]
```diff
@@ -236,6 +243,8 @@ static uint8_t GBSIOLockstepNodeWriteSC(struct GBSIODriver* driver, uint8_t valu
 			mTimingDeschedule(&driver->p->p->timing, &driver->p->event);
 			mTimingDeschedule(&driver->p->p->timing, &node->event);
 			mTimingSchedule(&driver->p->p->timing, &node->event, 0);
+		} else {
+			mLOG(GB_SIO, FATAL, "GBSIOLockstepNodeWriteSC() failed to write to masterClaimed\n");
```
Comment (endrift) [inline src/gb/sio/lockstep.c]: This should absolutely not be `FATAL`.

File: src/gba/sio/lockstep.c [hunk 6]
```diff
@@ -8,7 +8,23 @@
 #include <mgba/internal/gba/gba.h>
 #include <mgba/internal/gba/io.h>
 
-#define LOCKSTEP_INCREMENT 3000
+#define LOCKSTEP_INCREMENT 128
```
Comment (endrift) [inline src/gba/sio/lockstep.c]: This sounds like it's going to be a massive overhead. This is way too low.

File: src/gba/sio/lockstep.c [hunk 7]
```diff
@@ -8,7 +8,23 @@
...
+
+#if MGBA_LOCK_STEP_USE_MUTEX
+#	define LOCK_LOCK_STEP(node) MutexLock(&node.mutex)
```
Comment (endrift) [inline src/gba/sio/lockstep.c]: Don't indent.

File: src/gba/sio/lockstep.c [hunk 8]
```diff
@@ -19,7 +35,14 @@ static uint16_t GBASIOLockstepNodeNormalWriteRegister(struct GBASIODriver* drive
 static void _GBASIOLockstepNodeProcessEvents(struct mTiming* timing, void* driver, uint32_t cyclesLate);
 
 void GBASIOLockstepInit(struct GBASIOLockstep* lockstep) {
-	mLockstepInit(&lockstep->d);
+	GBASIOLockstepInit2(lockstep, true);
```
Comment (endrift) [inline src/gba/sio/lockstep.c]: Why is this needed?

File: src/gba/sio/lockstep.c (hunk 7 above)
Comment (endrift) [reply src/gba/sio/lockstep.c]: Also lockstep is one word.

File: src/gba/sio/lockstep.c [hunk 9]
```diff
@@ -148,13 +190,17 @@ static uint16_t GBASIOLockstepNodeMultiWriteRegister(struct GBASIODriver* driver
...
+	UNLOCK_LOCK_STEP(node->p->d);
+
 	return value;
 }
 
 static void _finishTransfer(struct GBASIOLockstepNode* node) {
 	if (node->transferFinished) {
 		return;
 	}
+
```
Comment (endrift) [inline src/gba/sio/lockstep.c]: Unnecessary code churn.

File: src/gba/sio/lockstep.c [hunk 10]
```diff
@@ -223,30 +279,30 @@ static int32_t _masterUpdate(struct GBASIOLockstepNode* node) {
 		node->p->multiRecv[2] = 0xFFFF;
 		node->p->multiRecv[3] = 0xFFFF;
 		needsToWait = true;
-		ATOMIC_STORE(node->p->d.transferActive, TRANSFER_STARTED);
-		node->nextEvent += 512;
+		SAFE_STORE(node->p->d.transferActive, TRANSFER_STARTED);
+		node->nextEvent += LOCKSTEP_INCREMENT;
```
Comment (endrift) [inline src/gba/sio/lockstep.c]: This was a separate increment for a reason--it's related to the minimum transfer duration.

File: src/gba/sio/lockstep.c [hunk 11]
```diff
@@ -269,16 +325,28 @@ static int32_t _masterUpdate(struct GBASIOLockstepNode* node) {
...
 	}
 	return node->nextEvent;
 }
 
 static uint32_t _slaveUpdate(struct GBASIOLockstepNode* node) {
-	node->d.p->multiplayerControl.ready = node->p->attachedMulti == node->p->d.attached;
+	LOCK_LOCK_STEP(node->p->d);
+
+	enum mLockstepPhase transferActive;
+	int attachedMulti, attached;
```
Comment (endrift) [inline src/gba/sio/lockstep.c]: Separate lines.

File: src/gba/sio/lockstep.c [hunk 12]
```diff
@@ -288,6 +356,8 @@ static uint32_t _slaveUpdate(struct GBASIOLockstepNode* node) {
 	case TRANSFER_FINISHING:
 		break;
 	case TRANSFER_STARTED:
+		if (node->p->d.unusedCycles(&node->p->d, node->id) > node->eventDiff)
```
Comment (endrift) [inline src/gba/sio/lockstep.c]: Braces.

File: src/gba/sio/lockstep.c [hunk 13]
```diff
@@ -315,6 +385,8 @@ static uint32_t _slaveUpdate(struct GBASIOLockstepNode* node) {
 		signal = true;
 		break;
 	case TRANSFER_FINISHED:
+		if (node->p->d.unusedCycles(&node->p->d, node->id) > node->eventDiff)
```
Comment (endrift) [inline src/gba/sio/lockstep.c]: Braces.

File: src/platform/qt/MultiplayerController.cpp [hunk 14]
```diff
@@ -16,43 +16,76 @@
 
 using namespace QGBA;
 
+#if MGBA_LOCK_STEP_USE_MUTEX
+#define ACQUIRE_CONTROLLER(controller, lockstep) MutexLock(&lockstep->mutex);
+#define RELEASE_CONTROLLER(controller, lockstep) MutexUnlock(&lockstep->mutex);
+#define CORE_THREAD_WAIT(player, lockstep) mCoreThreadWaitFromThread(player->controller->thread());
+#define CORE_THREAD_STOP_WAITING(player, lockstep) mCoreThreadStopWaiting(player->controller->thread());
+
```
Comment (endrift) [inline src/platform/qt/MultiplayerController.cpp]: Unnecessary newline.

File: src/platform/qt/MultiplayerController.cpp [hunk 15]
```diff
@@ -16,43 +16,76 @@
...
+#else // MGBA_LOCK_STEP_USE_MUTEX
+#define ACQUIRE_CONTROLLER(controller, lockstep) controller->m_lock.lock()
+#define RELEASE_CONTROLLER(controller, lockstep) controller->m_lock.unlock()
+#define CORE_THREAD_WAIT(player, lockstep) mCoreThreadWaitFromThread(player->controller->thread());
```
Comment (endrift) [inline src/platform/qt/MultiplayerController.cpp]: Why #define this if it's the same on both sides of the conditional?

File: src/platform/qt/MultiplayerController.cpp [hunk 16]
```diff
@@ -16,43 +16,76 @@
...
+#define CORE_THREAD_STOP_WAITING(player, lockstep) mCoreThreadStopWaiting(player->controller->thread());
+#endif // MGBA_LOCK_STEP_USE_MUTEX
+
+MultiplayerController::Player::Player(
```
Comment (endrift) [inline src/platform/qt/MultiplayerController.cpp]: Please match existing style.

File: src/platform/qt/MultiplayerController.h [hunk 17]
```diff
@@ -42,6 +45,17 @@ Q_OBJECT
 
 private:
 	struct Player {
+		Player(
```
Comment (endrift) [inline src/platform/qt/MultiplayerController.h]: Match existing style.

File: src/util/ring-fifo.c [hunk 18]
```diff
@@ -7,6 +7,18 @@
 
 #include <mgba-util/memory.h>
 
+#if defined _MSC_VER
```
Comment (endrift) [inline src/util/ring-fifo.c]: Why is this needed? It should already include the file these are defined in.

File: include/mgba-util/platform/posix/threading.h (hunk 2 above)
Comment (kakashidinho) [reply include/mgba-util/platform/posix/threading.h]: In your callback code in `MultilplayerController` you have another lock. I noticed it is unnecessary if mutex is used with lockstep anw. So make it recursive. Anw, the lock call in callback might not be needed. So the recursive mutex is not really neccessary.
For the platform porting, right now i only see win32 and pthread ans they are already ported

File: include/mgba/core/lockstep.h (hunk 3 above)
Comment (kakashidinho) [reply include/mgba/core/lockstep.h]: Win32 atomic function does not support abitrary type

**Comment by kakashidinho** [comment]
I tested pokemon ruby and even though it can play battle and trade now. After exitting the multiplayer mode, the link errors still occurred sometimes. Pokemon emerald seems fine.
So this might not be the final fix for this PR.

**Comment by kakashidinho** [comment]
For the mutex approach. As mentioned, your old code used mutex in the callbacks and they are in turn invoked in every lockstep event processing, if there is any overhead, it already exists. So just by introducing mutex to the event processing, the atomic ops will not be needed anymore (and possibly removing mutex from callbacks).

File: src/util/ring-fifo.c (hunk 18 above)
Comment (kakashidinho) [reply src/util/ring-fifo.c]: Again, due to rigid nature of win32 interlocked functions. Pointers need different versions of atomic function, but in this module, the code assumes the atomic* version for integers can work for pointers. So I have to do this. It's not best way, but it the quick way without modifying the resting the code to use something like `ATOMIC_STORE_PTR` for example.

File: src/gba/sio/lockstep.c (hunk 8 above)
Comment (kakashidinho) [reply src/gba/sio/lockstep.c]: To allow caller to choose to avoid init the base struct again, in case he/she wants to defer the initialization of derived struct such as the code in `MultiplayerController` which calls `mLockstepInit ` in constructor and separately calls `GBASIOLockstepInit ` in `attachGame`

**Comment by kakashidinho** [comment]
Close due to link error still occurs occasionally

**Comment by endrift** [comment]
I merged your changes with some stylistic (mostly) fixes on my end.
